### DISCUSSION TITLE
Keeping TS types when wrapping mapState function

### DISCUSSION THREAD

**Discussion by agualis** [vuejs/pinia#606] [Help and Questions] [answered]
We are incrementally migrating a big `Vue 2 JS` project from `Vuex` to `Pinia` where we had many usages of `mapGetters/mapState`.

Instead of directly using, for instance:

```javascript
// Vue SFC component 
import {mapState} from 'pinia';
import {useUserStore} from 'user-store';

// Options API 
...mapState(useUserStore,  {loggedIn: 'loggedIn'});
```

We would like to wrap all/some user state functions into something like: 

```javascript
// User pinia store (user-store.js)
import {mapState} from 'pinia';
import {useUserStore} from 'user-store';

export const mapUserState = (args) => mapState(useUserStore, args);
```

```javascript
// Vue SFC component 
import {mapUserState} from 'user-store';

// Options API 
...mapUserState({loggedIn: 'loggedIn'});
```  

The wrapper works but we are losing TS type inference. 

Here is a super simple project showing the issue: 

https://github.com/agualis/pinia-wrappers

Using plain `mapState` in VSCODE resolves `this.loggedIn` (boolean): 
![without-wrapping](https://user-images.githubusercontent.com/1316240/128845707-8cc1b5aa-4b44-4716-b586-0a9a6547895a.png)

Using custom `mapUserState` in VSCODE does not resolve `this.loggedIn` (any): 

![wrapped](https://user-images.githubusercontent.com/1316240/128845768-b306f925-7a48-4505-ac4a-1841ef92c2a6.png)

This is the type hint that I see in the wrapped version. It seems correct but not 100% sure:

![type-hint](https://user-images.githubusercontent.com/1316240/128849310-abbf1ff9-accf-49b8-8c6c-c965732e4b1f.png)

Does anyone has any clue on why this is happening and how to make it work?   Could it be just an unrelated Vetur issue? 🙏🏽

The same happens when passing array params instead of object ones to mapState: 

javascript
```
   ...mapState(useUserStore,  ['loggedIn']),
```

![array-params](https://user-images.githubusercontent.com/1316240/128850510-8a3e4f5f-df4d-46ad-a3ab-e6f58326f0d5.png)

Is it a known drawback of `pinia` types?

**Comment by posva** [comment]
for the array, you need to pass `['loggedIn'] as const` so TS infer it as a tuple instead of `string[]`

About you helper, you likely need to add a generic to it, something in the lines of 

```js
function mapHelper<T>(mapper: T) {
  return mapState(useUserStore, mapper)
}
```

**Reply by agualis** [reply]
The problem is that we are not using TS but `JS + JSDOC  types`. I'll try to get the same thing by using JSDOC generics. 

Thanks for your answer 🙌

**Reply by posva** [reply, marked as answer]
Ah yeah, then

```js
/**
 * @function
 * @template T
 * @param {T} mapper
 */

const mapHelper = (mapper) => useState(useStore, mapper)
```

You can also create a `d.ts` where you create the type in TS and then import it:

```js
/** @type {import('./types.d.ts').ExportedType} */
const mapHelper = (mapper) => useState(useStore, mapper)
```

**Reply by agualis** [reply]
Thanks a lot for your answers, very appreciated 🙌

> for the array, you need to pass ['loggedIn'] as const so TS infer it as a tuple instead of string[]

In JSDoc there is no `as const` but I used a [workaround](https://github.com/agualis/pinia-wrappers/blob/d7f24d440bb06364a53b70276c163e8da190f5df/src/App.vue#L26) that infers tuple as `['loggedIn']` but it is not working 😓

I had also tried your suggested [`@template` solution](https://github.com/agualis/pinia-wrappers/blob/d7f24d440bb06364a53b70276c163e8da190f5df/src/stores/user-store.js#L25) without success.

Very happy with the amazing pinia DX anyways. Keep on rocking!

**Reply by posva** [reply]
Try creating the type in a `d.ts` file then so you don't have type restrictions. T probably needs to extends a type though 🤔 Take a look at the types on the mapHelpers.ts file, you should be able to import some of them
